### PR TITLE
[native] Replace discoveryAddressLookup by CoordinatorDiscoverer

### PR DESCRIPTION
**Pull request by tanjialiang**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/Announcer.cpp [hunk 1]
```diff
@@ -125,7 +125,7 @@ void Announcer::makeAnnouncement() {
   }
 
   try {
-    auto newAddress = discoveryAddressLookup_();
+    auto newAddress = coordinatorDiscoverer_->fetchCoordinatorAddress();
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/Announcer.cpp:128]:
```
nit: s/fetchCoordinatorAddress/lookup/ a shorter name?
```

File: presto-native-execution/presto_cpp/main/Announcer.cpp [hunk 2]
```diff
@@ -74,7 +74,7 @@ Announcer::Announcer(
     const std::string& address,
     bool useHttps,
     int port,
-    std::function<folly::SocketAddress()> discoveryAddressLookup,
+    const std::shared_ptr<CoordinatorDiscoverer> coordinatorDiscoverer,
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/Announcer.cpp:77]:
```
const std::shared_ptr<CoordinatorDiscoverer>&

std::shared_ptr<CoordinatorDiscoverer> -> std::move() in ctor intializer
```

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#include <folly/SocketAddress.h>
+
+namespace facebook::presto {
+
+class CoordinatorDiscoverer {
+ public:
+  /// Fetch and update the cached coordinator address. Returns the newly
+  /// fetched address if success, otherwise the previously cached address. The
+  /// return address will be an empty address if fetch never succeeded.
+  virtual folly::SocketAddress fetchCoordinatorAddress();
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h:25]:
```
/// Fetches the coordinator address and update the cached address. The function returns the refreshed one on fetch success or the previously cached address. It only returns empty address if the fetches have never succeeded.
s/fetchCoordinatorAddress/update/
```

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  /// Returns the cached coordinator address.
+  folly::SocketAddress getCachedCoordinatorAddress();
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h:28]:
```
s/getCachedCoordinatorAddress/get/
```

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#include <folly/Uri.h>
+#include "presto_cpp/main/common/Configs.h"
+
+namespace facebook::presto {
+
+folly::SocketAddress CoordinatorDiscoverer::fetchCoordinatorAddress() {
+  // Check if discovery URI is specified. Presto-on-Spark doesn't specify it.
+  auto discoveryUri = SystemConfig::instance()->discoveryUri();
+  if (!discoveryUri.has_value()) {
+    // Returning the previous address. Empty address if fetching never
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.cpp:25]:
```
    // Returns the previous cached address or empty address if fetch never succeeded.
```

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+#include "presto_cpp/main/CoordinatorDiscoverer.h"
+#include <folly/Uri.h>
+#include "presto_cpp/main/common/Configs.h"
+
+namespace facebook::presto {
+
+folly::SocketAddress CoordinatorDiscoverer::fetchCoordinatorAddress() {
+  // Check if discovery URI is specified. Presto-on-Spark doesn't specify it.
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.cpp:22]:
```
s/specify it./specifies it./
```

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 7]
```diff
@@ -517,26 +518,8 @@ void PrestoServer::stop() {
...
-  // Check if discovery URI is specified. Presto-on-Spark doesn't specify it.
-  auto discoveryUri = SystemConfig::instance()->discoveryUri();
-  if (!discoveryUri.has_value()) {
-    PRESTO_STARTUP_LOG(INFO)
-        << "Discovery URI is not specified - will not run Announcer.";
-    return nullptr;
-  }
-
-  try {
-    auto uri = folly::Uri(discoveryUri.value());
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:530]: Shall we keep this failure handling?

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <folly/SocketAddress.h>
+
+namespace facebook::presto {
+
+class CoordinatorDiscoverer {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h:20]: It might be good to have unit test for this new class to test separately?

**Review by xiaoxmeng — COMMENTED** [review]
@tanjialiang nice change! The code is now more organized! Some minors % request for test on new class. Thanks!

File: presto-native-execution/presto_cpp/main/Announcer.cpp (hunk 2 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/Announcer.cpp:77]: std::move() has no effect in a const variable. We need a copy anyways. I can change the parameter type to be const &, to reduce one copy. That been said, I think compiler shall do the copy optimization regardless.

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.cpp (hunk 6 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.cpp:22]: doesn't specify it

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 7 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:530]: Discussed offline

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ public:
+  /// Fetches the coordinator address and update the cached address. The
+  /// function returns the refreshed one on fetch success or the previously
+  /// cached address. It will return an empty address if the fetches have never
+  /// succeeded.
+  virtual folly::SocketAddress lookup();
+
+  /// Returns the cached coordinator address.
+  folly::SocketAddress getCached();
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h:29]:
```
nit: s/getCached/cached/
```

**Review by xiaoxmeng — APPROVED** [review]
@tanjialiang Looks great. Thanks!

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h (hunk 9 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h:29]: I feel getCached is more readable, referring to other places in the codebase.

Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.h:29]: Shall we let PeriodicTaskManager to update coordinator address and both annoucement and http filter simply get the latest address? Then this object only expose update() and get() to methods?

File: presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.cpp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+folly::SocketAddress CoordinatorDiscoverer::lookup() {
+  // Check if discovery URI is specified. Presto-on-Spark doesn't specify it.
+  auto discoveryUri = SystemConfig::instance()->discoveryUri();
+  if (!discoveryUri.has_value()) {
+    // Returns the previous cached address or empty address if fetch never
+    // succeeded.
+    return coordinatorAddress_;
+  }
+
+  auto uri = folly::Uri(discoveryUri.value());
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/CoordinatorDiscoverer.cpp:30]: Shall we make the access to coordinatorAddress_ thread safe? Also print a log whenever we detect a coordinatorAddress_ change? Thanks!

**Pull request merged**